src/cursor/cur_std.c

```diff
@@ -710,12 +710,19 @@ __wt_cursor_cache(WT_CURSOR *cursor, WT_DATA_HANDLE *dhandle)
      */
     __wt_cursor_bound_reset(cursor);
 
-    /* Don't keep buffers allocated for cached cursors. */
-    __wt_buf_free(session, &cursor->key);
-    __wt_buf_free(session, &cursor->value);
-
-    /* Discard the underlying WT_CURSOR_BTREE buffers. */
-    __wt_btcur_cache((WT_CURSOR_BTREE *)cursor);
+    /*
+     * Check to see if the top level key/value buffers in the cursor are allocated. If so, we want
+     * to keep them around, they may be useful for the next use of this cursor. They will be freed
+     * by the session cursor sweep. If they are not allocated, free all the buffers in the cursor.
+     */
+    if (__wt_cursor_has_cached_memory(cursor)) {
+        F_SET(cursor, WT_CURSTD_CACHED_WITH_MEM);
+        cursor->key.data = NULL;
+        cursor->key.size = 0;
+        cursor->value.data = NULL;
+        cursor->value.size = 0;
```

Comment:
For re-reviewer(s): I added these lines, this will ensure that anyone opening a cursor, and getting a cached one, will not see anything in the key or value.

Reply:
It feels like this should be in `include/buf_inline.h` but looking at existing code there are about a dozen places where the code resets the `data` and `size` fields of a WT_ITEM. Should we consider a cleanup ticket for that? I'm on the fence. Something like `wt_buf_reset`?

Reply:
I think it's a good idea, I've opened WT-12898.